compiler/lib/code.ml

```diff
@@ -358,14 +358,19 @@ module Native_string = struct
     | Utf _, Byte _ | Byte _, Utf _ -> false
 end
 
+type int_kind =
+  | Regular
+  | Int32
+  | Native
+
 type constant =
   | String of string
   | NativeString of Native_string.t
   | Float of float
   | Float_array of float array
   | Int64 of int64
   | Tuple of int * constant array * array_or_not
-  | Int of int32
+  | Int of int_kind * int32
```

Comment:
Have you considered just adding two new constructors: Int32 and NativeInt to constant ? I feel the end result would be nicer and less error prone.

Reply:
That might be nicer indeed!

So, we need to change `Parse_bytecode.Constants.parse` and `Ocaml_compiler.constant_of_const` so that when producing JavaScript code, we convert `int32` and `nativeint` into regular `int` early and then only deal with regular integer. Then, functions `int_bin_op` and `shift` only have to deal with regular integers.

@OlivierNicole Do you want to give it a try?

Reply:
We can have a commit that just adds these two constructors here. And, then, the target specific code would be in #1655.

Reply:
> @OlivierNicole Do you want to give it a try?

@vouillon Sure, I’ll do it.

Reply:
@hhugo I have modified the relevant commit to implement your suggestion.